### PR TITLE
Adding login tests and related classes and implementations

### PR DESCRIPTION
**Pull request by PhysicsJon**

### CODE REVIEW

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Authorization/Authorization.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        public Authorization(UITest test, ScenarioContext context)
+        {
+            _test = test;
+            _context = context;
+        }
+        
+        [Given(@"that a User is not logged in")]
+        public void GivenThatAUserIsNotLoggedIn()
+        {
+            var builder = new UriBuilder(_test.Driver.Url)
```
Comment (talawson05) [inline src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Authorization/Authorization.cs]: should this bit be moved out somewhere, can called like OpenLoginPage or something?
Also what happens when we're not running on windows?

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/BrowseSolutions/DownloadAttachmentSteps.cs [hunk 2]
```diff
@@ -68,7 +67,7 @@ public async Task GivenAnAttachmentHasBeenProvidedForTheSolution(string s1, stri
 
         [Given(
             @"(a|an) (Roadmap|NHS Assured Integrations|Authority Provided Solution Document) attachment has not been provided for the Solution")]
-        public void GivenAnIntegrationsAttachmentHasNotBeenProvidedForTheSolution(string s1, string s2)
+        public void GivenAnIntegrationsAttachmentHasNotBeenProvidedForTheSolution(string _1, string _2)
```
Comment (talawson05) [inline src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/BrowseSolutions/DownloadAttachmentSteps.cs:70]: is this a resharper thing?

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Hooks.cs [hunk 3]
```diff
@@ -19,6 +20,11 @@ public Hooks(UITest test, ScenarioContext context)
         [AfterScenario]
         public async Task AfterScenario()
         {
+            var screenshot = ((ITakesScreenshot) _test.Driver).GetScreenshot();
+            var fileName = _context.ScenarioInfo.Title;
+
+            screenshot.SaveAsFile(fileName + ".png", ScreenshotImageFormat.Png);
```
Comment (talawson05) [inline src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Hooks.cs]: are these going into a folder or just where we're running them from?
Is it worth having a config flag to turn this on and off?
How about a techdebt ticket to attach the screenshots to the reports in AzDo?

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Utils/EnvironmentVariables.cs [hunk 4]
```diff
@@ -60,8 +69,23 @@ private static string JsonConfigValues(string section, string defaultValue)
...
+            var uri = "https://docker.for.win.localhost";
+            //if (Debugger.IsAttached)
+            //{
+                //var builder = new UriBuilder(uri)
+                //{
+                //    Host = "localhost"
+                //};
+
+                //uri = builder.Uri.ToString();
+            //}
```
Comment (talawson05) [inline src/NHSDPublicBrowseAcceptanceTests.Tests/Utils/EnvironmentVariables.cs]: is this debugger localhost no longer needed?

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/BrowseSolutions/DownloadAttachmentSteps.cs (hunk 2 above)
Comment (PhysicsJon) [reply src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/BrowseSolutions/DownloadAttachmentSteps.cs:70]: yes

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Hooks.cs (hunk 3 above)
Comment (PhysicsJon) [reply src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Hooks.cs]: I left this in as i'd been debugging, will be removing it shortly

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Utils/EnvironmentVariables.cs (hunk 4 above)
Comment (PhysicsJon) [reply src/NHSDPublicBrowseAcceptanceTests.Tests/Utils/EnvironmentVariables.cs]: I forgot I had commented out, will be removing the comments shortly

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Utils/EnvironmentVariables.cs [hunk 5]
```diff
@@ -60,8 +70,23 @@ private static string JsonConfigValues(string section, string defaultValue)
 
             return string.IsNullOrEmpty(result) ? defaultValue : result;
         }
-    }
 
+        private static string DefaultUri()
+        {
+            var uri = "https://docker.for.win.localhost";
```
Comment (talawson05) [inline src/NHSDPublicBrowseAcceptanceTests.Tests/Utils/EnvironmentVariables.cs]: for me, this doesn't work with https
If i use http then it works fine
I also don't understand why you need to be playing with the host?

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Authorization/Authorization.cs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            {
+                Host = "docker.for.win.localhost"
+            };
+            _test.Driver.Navigate().GoToUrl(builder.Uri.ToString());
```
Comment (talawson05) [inline src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Authorization/Authorization.cs]: when the test first starts it lands on publicbrowse homepage.

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Utils/EnvironmentVariables.cs (hunk 5 above)
Comment (PhysicsJon) [reply src/NHSDPublicBrowseAcceptanceTests.Tests/Utils/EnvironmentVariables.cs]: all of that will be removed soon (when i have finalised the configuration)

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Authorization/Authorization.cs (hunk 1 above)
Comment (PhysicsJon) [reply src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Authorization/Authorization.cs]: This is getting removed altogether in the next commit

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Utils/EnvironmentVariables.cs (hunk 5 above)
Comment (PhysicsJon) [reply src/NHSDPublicBrowseAcceptanceTests.Tests/Utils/EnvironmentVariables.cs]: URI is changing in the next commit

File: src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Authorization/Authorization.cs (hunk 6 above)
Comment (PhysicsJon) [reply src/NHSDPublicBrowseAcceptanceTests.Tests/Steps/Authorization/Authorization.cs]: This is being removed altogether in the next commit

**Review by talawson05 — APPROVED** [review]

**Pull request merged**